control/state.py

```diff
@@ -222,6 +227,49 @@ def get_omap_version(self) -> int:
                 f" invalid number of values ({value_list}).")
             raise
 
+    def lock_omap(self, duration):
+        got_lock = False
+
+        for i in range(1, self.omap_file_lock_retries):
+            try:
+                with rados.WriteOpCtx() as write_op:
+                    self.ioctx.lock_exclusive(self.omap_name, self.OMAP_FILE_LOCK_NAME, self.OMAP_FILE_LOCK_COOKIE, "OMAP file changes lock", duration, 0)
+                    got_lock = True
+                    break
+            except rados.ObjectExists as ex:
+                self.logger.info(f"We already locked the OMAP file")
+                got_lock = True
+                break
+            except rados.ObjectBusy as ex:
+                self.logger.warning(f"Someone else locked the OMAP file, will try again in {self.omap_file_lock_retry_sleep_interval} seconds")
+                time.sleep(self.omap_file_lock_retry_sleep_interval)
+            except Exception as ex:
+                self.logger.error(f"Unable to lock OMAP file: {ex}. Exiting!")
+                raise
+
+        if not got_lock:
+            self.logger.error(f"Unable to lock OMAP file after {self.omap_file_lock_retries} retries. Exiting!")
+            raise Exception("Unable to lock OMAP file")
+
+        omap_version = self.get_omap_version()
+
+        if omap_version > self.version:
+            self.logger.warning(f"Local version {self.version} differs from OMAP file version {omap_version}, need to read the OMAP file")
+            self.unlock_omap()
+            raise OSError(errno.EAGAIN, "Unable to lock OMAP file, file not current", self.omap_name)
+
+        return True
+
+    def unlock_omap(self):
+        try:
+            with rados.WriteOpCtx() as write_op:
```

Comment:
@gbregman, Not sure about:
```python
          with rados.WriteOpCtx() as write_op:
                self.ioctx.unlock(self.omap_name, self.OMAP_FILE_LOCK_NAME, self.OMAP_FILE_LOCK_COOKIE)
```

`write_op` is not referenced while locking/unlocking. 

- Do we need to create write context for using ioctx.lock/unlock?
- Could we combine locking and omap_cmp operation for version to get "locked and synced" invariant?

@idryomov, @oritwas  could you share your thoughts?

Reply:
@baum as I am not that familiar with using the Rados functions I just followed the examples I saw.